### ISSUE TITLE
[codemod] Unused params Causes next build Failure After Async API Codemod Execution

### ISSUE THREAD

**Issue by CastaChick** [vercel/next.js#73296]
### Link to the code that reproduces this issue

https://github.com/CastaChick/next_codemod_example

### To Reproduce

1. Create new next app using `next@14` by `npx create-next-app@14` command.
2. Add new pages with/without unused params

`app/example-unused-params/[id]/page.tsx`
```tsx
interface ExampleUnusedParamsProps {
   params: {
        id: string;
   }
}

// eslint-disable-next-line @typescript-eslint/no-unused-vars
export default function ExampleUnusedParams({params}: ExampleUnusedParamsProps) {
  return (
    <div>foo</div>
  )
}
```
`app/example-used-params/[id]/page.tsx`
```tsx
interface ExampleUsedParamsProps {
   params: {
        id: string;
   }
}

export default function ExampleUsedParams({params}: ExampleUsedParamsProps) {
  console.log(params.id)
  return (
    <div>foo</div>
  )
}
```

3. Run the migration command to `next@15` by running `npx @next/codemod@canary upgrade latest`
4. Then build the next app by running `npm run build` ← build failed

```
src/app/example-unused-params/[id]/page.tsx
Type error: Type 'ExampleUnusedParamsProps' does not satisfy the constraint 'PageProps'.
  Types of property 'params' are incompatible.
    Type '{ id: string; }' is missing the following properties from type 'Promise<any>': then, catch, finally, [Symbol.toStringTag]
```

### Current vs. Expected behavior

In Pull Request #71664 , it appears that transform is not executed when `params` is not accessed within a function. However, even if `params` is not used within the function, the build process (`next build`) will fail if `params` is not wrapped in a `Promise`.

To ensure developers can confidently build their projects after running `npx @next/codemod@canary next-async-request-api .`, I suggest either removing this behavior or ignoring unused, synchronously declared `params` during the build process. This adjustment would provide a smoother and more reliable development experience.

### Provide environment information

```bash
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.6.0: Mon Jul 29 21:14:30 PDT 2024; root:xnu-10063.141.2~1/RELEASE_ARM64_T6030
  Available memory (MB): 36864
  Available CPU cores: 12
Binaries:
  Node: 20.10.0
  npm: 10.2.3
  Yarn: 1.22.22
  pnpm: N/A
Relevant Packages:
  next: 15.0.3 // Latest available version is detected (15.0.3).
  eslint-config-next: 15.0.3
  react: 19.0.0-rc-66855b96-20241106
  react-dom: 19.0.0-rc-66855b96-20241106
  typescript: 5.7.2
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

Developer Experience, Output (export/standalone)

### Which stage(s) are affected? (Select all that apply)

next build (local)

**Comment by Innocent6303** [comment]
### Regarding the Suggested Behavior Change:

I completely understand the desire for a smoother and more reliable development experience, especially when running `next build`. However, I believe the behavior you’re suggesting — to ignore unused, synchronously declared `params` during the build process — might create more issues down the road.

The reason why TypeScript flags unused props, like `params`, is to ensure type safety during development. TypeScript helps developers catch potential errors early by reminding us when props are declared but not used. Ignoring or bypassing this in the build process could create blind spots in our code.

While you might feel that ignoring the unused `params` would allow the build to pass smoothly, it could lead to issues later. If Next.js were to make changes and, for some reason, `params` didn’t come through as expected, the application might break without any early warning. By enforcing this behavior in the build process, TypeScript helps to ensure that everything is in order before it’s too late.

So, while it might feel inconvenient to fix this during development, it’s actually TypeScript looking out for us, preventing potential issues from escalating when the app is running in production.

In short, this behavior ensures that props like `params` are always validated and accounted for, providing a more reliable and safe application.

If you would prefer to avoid this TypeScript warning, you could choose to use JSX instead of TypeScript. JSX won't enforce type checking, so it won't raise warnings about unused props. However, keep in mind that this might lead to issues down the line if you declare `params` but forget to provide it for any reason.

**Comment by CastaChick** [comment]
@Innocent6303 I completely agree with your opinion. Ignoring issues during the build process can lead to problems later on. I believe the best solution to increase the likelihood of a successful `next build` after running codemod is to wrap unused `params` in a `Promise` during the transformation process.

**Comment by CastaChick** [comment]
I believe the responsibility of a `codemod` is to recover from breaking changes introduced by updates and help ensure the application continues to function correctly. Regarding **warning-level** issues that existed before the update (e.g., `unused arguments`), I think it is preferable for the `codemod` to transform them rather than contextually handling or differentiating them within the transformation logic.
